tls/s2n_auth_selection.c

```diff
@@ -0,0 +1,195 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crypto/s2n_signature.h"
+#include "crypto/s2n_certificate.h"
+
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_kex.h"
+#include "tls/s2n_auth_selection.h"
+
+#include "utils/s2n_safety.h"
+
+/* This module should contain any logic related to choosing a valid combination of
+ * signature algorithm, authentication method, and certificate to use for authentication.
+ *
+ * We choose our auth methods by:
+ * 1. Finding a cipher suite with an auth method that we have valid certs for. In TLS1.3,
+ *    this is a no-op -- cipher suites do not specify an auth method.
+ * 2. Choosing a signature algorithm that matches both the auth method (if set) and the
+ *    available certs.
+ * 3. Selecting the cert that matches the chosen signature algorithm.
+ *
+ * This is a break from the original s2n pre-TLS1.3 flow, when we could choose certs and
+ * ciphers at the same time. Our cipher suites differentiate between "RSA" and "ECDSA",
+ * but not between "RSA" and "RSA-PSS". To make that decision, we need to wait until
+ * we've chosen a signature algorithm. This allows us to use RSA-PSS with existing
+ * TLS1.2 cipher suites.
+ */
+
+static int s2n_get_auth_method_for_cert_type(s2n_pkey_type cert_type, s2n_authentication_method *auth_method) {
+    switch(cert_type) {
+        case S2N_PKEY_TYPE_RSA:
+        case S2N_PKEY_TYPE_RSA_PSS:
+            *auth_method = S2N_AUTHENTICATION_RSA;
+            return S2N_SUCCESS;
+        case S2N_PKEY_TYPE_ECDSA:
+            *auth_method = S2N_AUTHENTICATION_ECDSA;
+            return S2N_SUCCESS;
+        default:
+            S2N_ERROR(S2N_ERR_CERT_TYPE_UNSUPPORTED);
+    }
+}
+
+static int s2n_get_cert_type_for_sig_alg(s2n_signature_algorithm sig_alg, s2n_pkey_type *cert_type) {
+    switch(sig_alg) {
+        case S2N_SIGNATURE_RSA_PSS_RSAE:
+        case S2N_SIGNATURE_RSA:
+            *cert_type = S2N_PKEY_TYPE_RSA;
+            return S2N_SUCCESS;
+        case S2N_SIGNATURE_ECDSA:
+            *cert_type = S2N_PKEY_TYPE_ECDSA;
+            return S2N_SUCCESS;
+        case S2N_SIGNATURE_RSA_PSS_PSS:
+            *cert_type = S2N_PKEY_TYPE_RSA_PSS;
+            return S2N_SUCCESS;
+        default:
+            S2N_ERROR(S2N_ERR_INVALID_SIGNATURE_ALGORITHM);
+    }
+}
+
+static int s2n_is_sig_alg_valid_for_cipher_suite(s2n_signature_algorithm sig_alg, struct s2n_cipher_suite *cipher_suite) {
+    s2n_pkey_type cert_type;
+    GUARD(s2n_get_cert_type_for_sig_alg(sig_alg, &cert_type));
+
+    /* RSA-PSS only supports Sign/Verify, and not Encrypt/Decrypt, which means that it MUST be used with an
+     * ephemeral Key Exchange Algorithm. */
+    if (cipher_suite->key_exchange_alg != NULL && !cipher_suite->key_exchange_alg->is_ephemeral) {
+        ne_check(cert_type, S2N_PKEY_TYPE_RSA_PSS);
+    }
```

Comment:
I'm torn here. Honestly, I wrote it that way (in reverse order) because of the convenience of ne_check. I can't return here, because there are potentially more checks to do, so to match the comment it would actually be:

```
if (cert_type == S2N_PKEY_TYPE_RSA_PSS) {
  notnull_check(cipher_suite->key_exchange_alg);
  if (!cipher_suite->key_exchange_alg->is_ephemeral) {
    return S2N_FAILURE; // And this should probably actually be a named error
  }
}
```

Which is definitely more verbose, but also probably clearer.

Reply:
i see, that makes sense.  and good point on the return early part.  The only thing with using the `ne_check` is now our error is `S2N_ERR_SAFETY` but as we are treating this function as a boolean, the actual error may not matter.